### PR TITLE
Add segindex to the posix share memory segment name

### PR DESCRIPTION
**Pull request by fanfuxiaoran**
For gp, there may be several segments running on the same host. So, if an application uses 'reloid' as the handle, such as pg_embedding, some concurrency issues will happen. That's due to the memory segment name collisions between those processes. Several different processes which have different segment ids, attach to the same shared memory segment as they have the same name "/PostgreSQL/reloid".

Resolve this issue by adding segindex to the name.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

**Comment by avamingli** [comment]
> For gp, there may be several segments running on the same host. So, if an application uses 'reloid' as the handle, such as pg_embedding, some concurrency issues will happen. That's due to the memory segment name collisions between those processes. Several different processes which have different segment ids, attach to the same shared memory segment as they have the same name "/PostgreSQL/reloid".

Woo, this is really a significant change with UPSTREAM.

I'm a little confused here, different segments on same host have their own  Postmaster process.
Their shared memory segment couldn't reach others.

> "/PostgreSQL/reloid"

Is it a absolute dir path for OS shared memory?

**Comment by avamingli** [comment]
> 
> > "/PostgreSQL/reloid"
> 
> Is it a absolute dir path for OS shared memory?

Have a look PG used shm_open() with the name directly, that may be a potential risk/bug of Postgres.
@fanfuxiaoran Maybe we could make a path for Postgres.

**Comment by fanfuxiaoran** [comment]
> > > "/PostgreSQL/reloid"
> > 
> > 
> > Is it a absolute dir path for OS shared memory?
> 
> Have a look PG used shm_open() with the name directly, that may be a potential risk/bug of Postgres. @fanfuxiaoran Maybe we could make a patch for Postgres:)

Actually, in dsm_impl.c, there are some comments about it:

File: src/backend/storage/ipc/dsm_impl.c [hunk 1]
```diff
@@ -219,7 +220,20 @@ dsm_impl_posix(dsm_op op, dsm_handle handle, Size request_size,
...
+	/*
+	 * Adding segindex to the name. As for gp, there may be several
+	 * segments running on the same host, which can lead to name
+	 * collisions between those processes.
+	 *
+	 * GpIdentity.segindex can only be set value when postmaster
+	 * starts, so it is safe to use it here.
+	 *
+	 * However, this increases the length of the name to exceed
+	 * its max available length. For instance, on maxos, the
```
Comment (avamingli) [inline src/backend/storage/ipc/dsm_impl.c:232]: `on maxos,` -> `macos`?

**Comment by fanfuxiaoran** [comment]
> > For gp, there may be several segments running on the same host. So, if an application uses 'reloid' as the handle, such as pg_embedding, some concurrency issues will happen. That's due to the memory segment name collisions between those processes. Several different processes which have different segment ids, attach to the same shared memory segment as they have the same name "/PostgreSQL/reloid".
> 
They named shared memory segments. Even different processes which have no parent-child relationship can have access to such a shared memory segment.
> Woo, this is really a significant change with UPSTREAM.
> 
> I'm a little confused here, different segments on same host have their own Postmaster process. Their shared memory segment couldn't reach others.
> 
> > "/PostgreSQL/reloid"
> 
> Is it a absolute dir path for OS shared memory?

Actually, I think it is the shared memory object name. On Linux, you can see them under directory /dev/shm.

**Comment by ashwinstar** [comment]
`"/PostgreSQL/reloid"` isn't this usage itself wrong as there is no guarantee would be able to create DSM with this handle even with single node PostgreSQL. `dsm_create()` and `dsm_postmaster_startup()` loop over `dsm_impl_op()` till able to get unused identifier. What if the identified (reloid) is used with random already, how does it deal with it.

**Comment by fanfuxiaoran** [comment]
> `"/PostgreSQL/reloid"` isn't this usage itself wrong as there is no guarantee would be able to create DSM with this handle even with single node PostgreSQL. `dsm_create()` and `dsm_postmaster_startup()` loop over `dsm_impl_op()` till able to get unused identifier. What if the identified (reloid) is used with random already, how does it deal with it.

Yes, I agree with that using `"/PostgreSQL/reloid"` is not a good idea.  
>What if the identified (reloid) is used with random already, how does it deal with it.

To avoid this issue, it doesn't use the reloid directly, instead, it makes it even by `reloid << 1`. Although there is still a risk there already is a handle generated by `random()` which is the same as the  `reloid << 1` , it can work most of the time.

We cannot make sure the applications use dsm handle correctly, and it is difficult for us to figure out such issues. To avoid struggling with it, I thought it may be a good idea to have a such pr.

**Comment by beeender** [comment]
A possible scenario: 
- QD creates a unique handle and dispatches it to QEs
- QE uses it to create/attach to DSM
- There are multiple segments on the same host, :boom: 
 
"QD creates a unique handle and dispatches it to QEs" This design sounds reasonable and it should be handled well.

File: src/backend/storage/ipc/dsm_impl.c [hunk 2]
```diff
@@ -219,7 +220,20 @@ dsm_impl_posix(dsm_op op, dsm_handle handle, Size request_size,
 	int			fd;
 	char	   *address;
 
-	snprintf(name, 64, "/PostgreSQL.%u", handle);
```
Comment (avamingli) [inline src/backend/storage/ipc/dsm_impl.c:222]:
```c
#if 0
  snprintf(name, 64, "/PostgreSQL.%u", handle);
#endif
```
Use #if 0 to keep with UPSTREAM as much as possible?

Comment (fanfuxiaoran) [reply src/backend/storage/ipc/dsm_impl.c:222]: Make sense

**Comment by fanfuxiaoran** [comment]
This pr can not totally resolve the conflict of the dsm id, so just close it.

**Pull request closed without merging**